docs/design/002-containerized-gadgets.md

```diff
@@ -0,0 +1,602 @@
+
+	for !gadget.Done() {
+		fmt.Print("event received: %+v\n", gadget.Next())
+	}
+}
+
+```
+
+## Reimplementing existing gadgets as containerized gadgets
+
+A long term vision of this support is to reimplement all gadgets as containerized gadgets, then
+deprecated the built-in ones. This section presents the features Inspektor Gadget should support for
+it and checks additional challenges by existing gadgets.
+
+### Features
+
+This section describes additional feature needed by containerized containers in Inspektor Gadget.
+Some of them are already supported and others are planned.
+
+#### Filtering and enrichment by mount namespace id
+
+Inspektor Gadget provides a set of helper functions in pkg/gadgets/common/mntns_filter.h for gadgets
+that want to filter and enrich events based on the mount namespace inode id information. Inspektor
+Gadget detects the presence of the `gadget_mntns_filter_map` map and populates it with the matching
+containers to filter. Enriching is done by looking for the presence of a field with `mnt_ns_id_t`
+type in the event.
+
+#### Endpoint enrichment
+
+Networking gadgets that want to enrich IP addresses with Pod and Services name can use the `gadget_l3endpoint_t` and `gadget_l4endpoint_t` types provided by Inspektor Gadget.
+
+NOTE: It's under development at the time of writing this document in
+https://github.com/inspektor-gadget/inspektor-gadget/pull/1825.
+
+#### `socket_filter` programs
+
+Attach ebpf programs of type `BPF_PROG_TYPE_SOCKET_FILTER`. TODO: What's special about these?
+
+#### Socket enricher
+
+Use the socket enricher from the ebpf code. This can be done in a very similar way to the
+`gadget_mntns_filter_map`: Inspektor Gadget looks for this map, automatically replaces and keeps it
+updated with the right one.
+
+#### Custom parameters
+
+Some gadgets define constants on the eBPF code to change its behavior. Those constants should be
+exposed to the user as parameters.
+
+It could be implemneted by using the pkg/params package.
+
+
+#### Enum convert
+
+Convert an enum to a string. Examples: pktTypeNames, qTypeNames, rCodeNames, etc. Some could use BTF
+from the kernel, but not all.
+
+Mauricio: I think we can use the BTF from the program too.
+
+#### String edit
+
+string manipulation routines. Example: parseLabelSequence to replace dns strings to dotted names.
+
+#### Bitfield
+
+Convert bitfield from an event to a human readable string. Example: bind's [optionsToString](https://github.com/inspektor-gadget/inspektor-gadget/blob/b57f2bae31a46b40d8e0204b85099ae37f15d21d/pkg/gadgets/trace/bind/tracer/tracer.go#L154).
+
+#### Iface index
+
+Convert network interface index to string. Example: bind's [BoundDevIf](https://github.com/inspektor-gadget/inspektor-gadget/blob/b57f2bae31a46b40d8e0204b85099ae37f15d21d/pkg/gadgets/trace/bind/tracer/tracer.go#L236-L250).
+
+#### Custom user space logic
+
+There are some gadgets that have a lot of logic in user space, to support this gadgets we'll need to
```

Comment:
I think we should tackle this soon. It's a very empowering feature and one that we already need.

Reply:
Right. I changed the wording to make it clear we'll do that later on.